WIP Use subpaths for hook mounts

oshoval opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Consider creating this PR as draft: https://github.com/kubevirt/kubevirt/blob/main/CONTRIBUTING.md#consider-opening-your-pull-request-as-draft
2. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

### What this PR does
Before this PR:

After this PR:

<!-- (optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*: -->
Fixes #

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: <!-- optional: slack, other GH issue, mailinglist, ... -->

### Special notes for your reviewer

<!-- optional -->

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
None
```

oshoval commented on pkg/virt-controller/services/template_test.go:
```diff
@@ -1078,7 +1078,7 @@ var _ = Describe("Template", func() {
 				Expect(pod.Spec.Containers[1].Image).To(Equal("some-image:v1"))
 				Expect(pod.Spec.Containers[1].ImagePullPolicy).To(Equal(k8sv1.PullPolicy("IfNotPresent")))
 				Expect(pod.Spec.Containers[1].VolumeMounts[0].MountPath).To(Equal(hooks.HookSocketsSharedDirectory))
-
+				Expect(pod.Spec.Containers[1].VolumeMounts[0].SubPath).To(Equal("0"))
```
tested already on pkg/hooks/manager_test.go

oshoval commented:
https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/14623/pull-kubevirt-unit-test-s390x/1919707114918383616
doesnt seem related to this PR, and also reminds this failure on other PR
https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/14520/pull-kubevirt-unit-test-arm64/1917208154534318080

/test pull-kubevirt-unit-test-s390x

oshoval commented:
/test pull-kubevirt-code-lint

orelmisan commented on docs/network/network-binding-plugin.md:
```diff
@@ -505,3 +505,10 @@ derived from virtual machines using the plugin.
 must be specified in the Kubevirt CR.
 See the user-guide network binding plugin [section](https://kubevirt.io/user-guide/network/network_binding_plugins/#register) on how to define it.
+
+## Network plugin custom sockets
```
Please consider splitting this change to a defecated commit.

orelmisan replied:
Please consider adjusting the wording to emphasize that KubeVirt scans the <dir_name> directory for Unix sockets to communicate with the sidecar.
This directory should not contain directly contain sockets used for other purposes.

orelmisan commented on pkg/hooks/manager_test.go:
```diff
@@ -25,6 +25,7 @@ import (
 	"net"
 	"os"
 	"path/filepath"
+	"strconv"
```
Will it make sense to split the changes to template.go from the changes to virt-launcher?

oshoval replied:
imho no need, as it is small
but i can take into account for next times

orelmisan reviewed (commented):
Thank you for the PR @oshoval.
Please see the inline comments.

oshoval commented on pkg/hooks/manager.go:
```diff
@@ -109,29 +109,30 @@ func (m *hookManager) collectSideCarSockets(numberOfRequestedHookSidecars uint,
 		}
 
 		for _, socket := range sockets {
-			select {
```
The latter, if there are no sockets at all, when you do expect sockets.

The function directly finishes if you expect no sockets, so this case isn't the important one, hence didn't see need to elaborate on it, will do if you still think we should.

orelmisan replied:
I suggest adjusting the commit message to explicitly state what had changed and the motivation behind it.

bgaussen commented on pkg/virt-controller/services/template.go:
```diff
@@ -868,10 +873,11 @@ func (t *templateService) newResourceRenderer(vmi *v1.VirtualMachineInstance, ne
 }
 
-func sidecarVolumeMount() k8sv1.VolumeMount {
+func sidecarVolumeMount(containetName string) k8sv1.VolumeMount {
```
Typo: containe**r**Name

oshoval commented:
Addressed comments (Thanks Benoit, Orel)

oshoval commented:
Updated 3rd commit message wrt upgrades, explaining why those are safe as discussed offline

oshoval commented:
/retest

orelmisan commented on pkg/virt-controller/services/template_test.go:
```diff
@@ -1078,7 +1078,7 @@ var _ = Describe("Template", func() {
 				Expect(pod.Spec.Containers[1].ImagePullPolicy).To(Equal(k8sv1.PullPolicy("IfNotPresent")))
 				Expect(pod.Spec.Containers[1].VolumeMounts[0].MountPath).To(Equal(hooks.HookSocketsSharedDirectory))
-
+				Expect(pod.Spec.Containers[1].VolumeMounts[0].SubPath).To(Equal("hook-sidecar-0"))
```
IIUC, this test tests node selectors.
Is there currently a dedicated test for a VMI with sidecars?

oshoval replied:
you are right, 
yes (unit test), in the same file, will update that one instead (+ will also add the env var UT you asked on other comment) there

orelmisan commented on pkg/hooks/hooks.go:
```diff
@@ -21,6 +21,7 @@ package hooks
 
 import (
 	"encoding/json"
+	"os"
```
Commit message: "virt-launcher, sidecars:"
The change is actually mostly in the VMI controller's template service.

nit: Please consider adding a link to k8s' `SubPath` documentation:
https://kubernetes.io/docs/concepts/storage/volumes/#using-subpath

orelmisan commented on pkg/virt-controller/services/template.go:
```diff
@@ -540,6 +540,11 @@ func (t *templateService) renderLaunchManifest(vmi *v1.VirtualMachineInstance, i
 					})
 			}
 		}
+		sidecarContainer.Env = []k8sv1.EnvVar{
+			k8sv1.EnvVar{
+				Name:  hooks.ContainerNameEnvVar,
+				Value: sidecarContainerName(i),
+			}}
```
Please consider moving this logic to `pkg/virt-controller/services/rendercontainer.go` as it already contains similar logic.

Please see `newSidecarContainerRenderer()` above.

Additionally, could it be covered by unit tests?

oshoval replied:
done + added it to ut

orelmisan reviewed (changes requested):
Thank you for the changes @oshoval.

oshoval commented on pkg/hooks/hooks.go:
```diff
@@ -65,3 +68,11 @@ func UnmarshalHookSidecarList(vmiObject *v1.VirtualMachineInstance) (HookSidecar
 	return hookSidecarList, nil
 }
+
+func GetSocketRootPath() string {
+	return HookSocketsSharedDirectory
+}
+
+func GetContainerName() string {
+	return os.Getenv(ContainerNameEnvVar)
+}
```
They are helper functions for sidecar developer,
for example for https://github.com/kubevirt/community/pull/294 developers
that will need to know the path on the compute container (which is the Join of those two) 
since they are so simple we can even drop them if desired

oshoval replied:
yes, this file, even without the new functions is the way to know the `HookSocketsSharedDirectory` path

oshoval commented on pkg/hooks/manager.go:
```diff
@@ -141,6 +133,23 @@ func (m *hookManager) collectSideCarSockets(numberOfRequestedHookSidecars uint,
 	return callbacksPerHookPoint, nil
 }
 
+func handleSidecarSocket(filePath string, callbacksPerHookPoint map[string][]*callBackClient) error {
```
For that, we need to deep copy the original
because if i change it in the function and return it, it still update the original one
I dont like the idea to deep copy the original just for that

orelmisan replied:
I meant something along the lines of:

```go
callbacksPerHookPoint := make(map[string][]*callBackClient)
...

CallbacksByHookName, err := handleSidecarSocket(filePath)
if err != nil {
...
}

maps.Copy(callbacksPerHookPoint, hookNameToCallbackClient)
```

orelmisan replied:
Lets keep as as is for now.

oshoval commented:
addressed comments https://github.com/kubevirt/kubevirt/compare/c2e969ff257fc7f624a0290154aea9b19150074b..1c005f76b7593565051eda6dc7c7043635590ebc

2nd push - rebase

oshoval commented:
reverted unrelated new line removal

oshoval commented:
/retest

oshoval commented:
/retest

orelmisan commented on pkg/virt-controller/services/template.go:
```diff
@@ -763,6 +763,8 @@ func newSidecarContainerRenderer(sidecarName string, vmiSpec *v1.VirtualMachineI
 		requestedHookSidecar.Image = os.Getenv(operatorutil.SidecarShimImageEnvName)
 	}
 
+	sidecarOpts = append(sidecarOpts, WithContainerName(sidecarName))
```
Any reason not to add the `WithContainerName()` option when initializing `sidecarOpts`?

orelmisan commented on pkg/virt-controller/services/template_test.go:
```diff
@@ -501,6 +501,12 @@ var _ = Describe("Template", func() {
 				Expect(pod.Spec.Containers[1].Name).To(Equal("hook-sidecar-0"))
 				Expect(pod.Spec.Containers[1].Image).To(Equal("some-image:v1"))
 				Expect(pod.Spec.Containers[1].ImagePullPolicy).To(Equal(k8sv1.PullPolicy("IfNotPresent")))
+				Expect(pod.Spec.Containers[1].VolumeMounts[0].SubPath).To(Equal("hook-sidecar-0"))
```
How about moving these assertions to `It("should call sidecar creators", func() {`? Since It is a dedicated sidecar test.

orelmisan commented on pkg/hooks/manager.go:
```diff
@@ -103,22 +103,38 @@ func (m *hookManager) collectSideCarSockets(numberOfRequestedHookSidecars uint,
 
-			processedSockets[socket.Name()] = true
+			for _, subEntry := range subEntries {
+				if subEntry.IsDir() {
```
Please consider also skipping non Unix socket files.

oshoval replied:
I prefer not in this PR, we can always do follow-ups upon needs,
it wasn't there before, and we can't encourage putting other files in this folder, as then someone can put non gRPC socket (fwiw also before this PR one couldn't put there other files)

orelmisan commented on pkg/hooks/manager_test.go:
```diff
@@ -129,27 +134,32 @@ var _ = Describe("HooksManager", func() {
 				defer os.Remove(socketPath)
 			}
 
 			manager := newManager(socketDir)
-			err := manager.Collect(uint(len(hookNameMap)), 10*time.Second)
+			err := manager.Collect(uint(len(hookNameList)), 10*time.Second)
 			Expect(err).ToNot(HaveOccurred())
 
 			callbackMaps := manager.CallbacksPerHookPoint
 
-			for _, hookPointName := range hookNameMap {
-				Expect(callbackMaps).Should(HaveKey(hookPointName))
-				Expect(callbackMaps[hookPointName]).Should(HaveLen(1))
+			for _, hook := range hookNameList {
+				Expect(callbackMaps).Should(HaveKey(hook.hookPointName))
+				Expect(callbackMaps[hook.hookPointName]).Should(HaveLen(1))
```
Can we compare the actual map to an expected map?
I think it would be easier to follow.

oshoval replied:
lets defer please, i prefer to keep the original subject on the PR, there is no regression, no need cosmetics in this PR

orelmisan reviewed (changes requested):
Thank you for the changes @oshoval

oshoval commented:
thx, addressed comments

orelmisan commented on pkg/virt-controller/services/template_test.go:
```diff
@@ -501,6 +501,7 @@ var _ = Describe("Template", func() {
 				Expect(pod.Spec.Containers[1].Name).To(Equal("hook-sidecar-0"))
 				Expect(pod.Spec.Containers[1].Image).To(Equal("some-image:v1"))
 				Expect(pod.Spec.Containers[1].ImagePullPolicy).To(Equal(k8sv1.PullPolicy("IfNotPresent")))
+
```
nit: Redundant newline.

orelmisan commented on docs/network/network-binding-plugin.md:
```diff
@@ -505,3 +505,10 @@ derived from virtual machines using the plugin.
+Some plugins may need to create custom sockets.
+Kubevirt scans `HookSocketsSharedDirectory` itself for gRPC sockets to communicate with the sidecars,
+hence this directory should not contain any other file.
+Due to this, please create a folder (i.e `userdata`) under sidecar's `HookSocketsSharedDirectory` directory and put there the required sockets.
```
same.

orelmisan commented on pkg/virt-controller/services/rendercontainer.go:
```diff
@@ -183,6 +183,15 @@ func WithReadinessProbe(vmi *v1.VirtualMachineInstance) Option {
 
+func WithContainerName(containerName string) Option {
+	return func(renderer *ContainerSpecRenderer) {
+		renderer.env = append(renderer.env, k8sv1.EnvVar{
+			Name:  hooks.ContainerNameEnvVar,
+			Value: containerName,
+		})
```
So hot about changing this function to be `WithEnvVars` to provide external environment variables?

oshoval replied:
i dont think it should be in this scope of this PR, we can do a follow-up if desired
it derails and will make the PR way less focused

orelmisan replied:
> i dont think it should be in this scope of this PR, we can do a follow-up if desired it derails and will make the PR way less focused It also makes the common part `envVars` to be a pattern, orthogonal to this PR, and maintainers might not like the idea (i dont like it fwiw), i suggest to refrain from it, at least on this PR please.
> 
> EDIT - alternative which is better and much simple is just add the env var to all containers, this is another benefit of follow-up, we can do the discussion in parallel

I don't follow how this suggestion can derail or make the PR less focused.
In any case the change should be negotiated with the relevant compute / core approvers.

orelmisan replied:
> can you write please exactly what you suggest about `WithEnvVars` and from there we advance ? (so we see we are in sync)

I suggest having a function, that will enable the caller to provide external environment variables:
```go
func WithEnvVars(envVars []k8sv1.EnvVar) Option {
	return func(renderer *ContainerSpecRenderer) {
		renderer.env = envVars
	}
}
```

It has greater extensibility potential over a dedicated function that just adds a specific env var.

oshoval replied:
problem is that we need to move it from Render, which is the common block now
unless you mean orthogonal for it, just to generalize `WithContainerName` which is other story ?
(in this case, easier to introduce that and indeed more focused, but i am not sure we need now something that generic)

oshoval commented on docs/network/network-binding-plugin.md:
```diff
@@ -505,3 +505,10 @@ derived from virtual machines using the plugin.
+## Network plugin custom sockets
+
+Some plugins may need to create custom sockets.
+Kubevirt scans `HookSocketsSharedDirectory` itself for gRPC sockets to communicate with the sidecars,
```
`/var/run/kubevirt-hooks` is not correct, because it is mounted as `/var/run/kubevirt-hooks/<container_name>`
due to this and since the file already use the const, it is better to keep it as const

oshoval commented:
Addressed comments

oshoval commented:
/hold

first commit extrracted to https://github.com/kubevirt/kubevirt/pull/14764
(will extract the container_name to its own PR, and will update once done)

oshoval commented:
Extracted CONTAINER_NAME feature to https://github.com/kubevirt/kubevirt/pull/14768
as discussed

basically this PR is ready to review, just please note first commit is cherry-picked, will update PR desc about it

vladikr commented:
@oshoval if this is related to a VEP , could you please add a reference to it in the description?

oshoval commented:
> @oshoval if this is related to a VEP , could you please add a reference to it in the description?

Hi Vladik
the follow-up is related to the VEP, and based on this one, there is link there
https://github.com/kubevirt/kubevirt/pull/14768
is it good enough ?
this PR is generic and improve isolation even if there was no VEP

orelmisan reviewed (approved):
Thank you @oshoval!

fossedihelm commented on pkg/hooks/manager.go:
```diff
@@ -113,19 +113,11 @@ func (m *hookManager) collectSideCarSockets(numberOfRequestedHookSidecars uint,
-				continue
-			} else if err != nil {
-				log.Log.Reason(err).Infof("Failed to process sidecar socket: %s", socket.Name())
+			err := handleSidecarSocket(filepath.Join(m.hookSocketSharedDirectory, socket.Name()), callbacksPerHookPoint)
```
This is not equivalent and could cause some issues.
Previously, in case of `notReady` we were skipping the `processedSockets[socket.Name()] = true` statement.
With this change no.
This could result in some issues since at line 112 we use `processedSockets[socket.Name()]` as a gate, and also check its len in the main for loop

fossedihelm replied:
I think you should change the `handleSidecarSocket` signature, returning the `notReady` too, and checking it:
```golang
				notReady, err := handleSidecarSocket(filepath.Join(subPath, subEntry.Name()), callbacksPerHookPoint)
				if notReady {
					continue
				}

				if err != nil {
					return nil, err
				}
```
`handleSidecarSocket` could be:
```golang
func handleSidecarSocket(filePath string, callbacksPerHookPoint map[string][]*callBackClient) (bool, error) {
	callBackClient, notReady, err := processSideCarSocket(filePath)
	if notReady {
		log.Log.Infof("Sidecar server might not be ready yet: %s", filePath)
		return true, nil
	}
	if err != nil {
		log.Log.Reason(err).Infof("Failed to process sidecar socket: %s", filePath)
		return false, err
	}

	for _, subscribedHookPoint := range callBackClient.subscribedHookPoints {
		callbacksPerHookPoint[subscribedHookPoint.GetName()] = append(callbacksPerHookPoint[subscribedHookPoint.GetName()], callBackClient)
	}

	return false, nil
}
```
or a more cryptic one:
```golang
func handleSidecarSocket(filePath string, callbacksPerHookPoint map[string][]*callBackClient) (bool, error) {
	callBackClient, notReady, err := processSideCarSocket(filePath)
	if notReady {
		log.Log.Infof("Sidecar server might not be ready yet: %s", filePath)
	}
	if err != nil {
		log.Log.Reason(err).Infof("Failed to process sidecar socket: %s", filePath)
	}

	for _, subscribedHookPoint := range callBackClient.subscribedHookPoints {
		callbacksPerHookPoint[subscribedHookPoint.GetName()] = append(callbacksPerHookPoint[subscribedHookPoint.GetName()], callBackClient)
	}

	return notReady, err
}
```

oshoval replied:
totally, my bad, the gate was in the function first and extracting that cause it
thanks nice find

fossedihelm replied:
> 2nd option isnt as good, best to fail fast, and not set (even without effect callbacksPerHookPoint in case of err / notReady)

That's why I said "cryptic one" :slightly_smiling_face:

fossedihelm replied:
Another (totally personal) opinion:
"I hate negated boolean".
`ready`  is better than `notReady`

But this requires a wider change that should be addressed out of this PR

oshoval replied:
thanks done, with slight diff (checking err before ready, because err it more important, even if the function should return only one of them enabled)

fossedihelm reviewed (commented):
> @fossedihelm can you take a look please ?

@oshoval Sure!
Thank you.

I don't have approval rights for this PR, so you need a root approver anyway :)

oshoval commented:
Thanks addressed comments

oshoval commented:
> I don't have approval rights for this PR, so you need a root approver anyway :)

hmm what do i miss ? (just curious of course)
https://github.com/kubevirt/kubevirt/blob/5f866200e20553d7a37886ad9de99feea1fffece/OWNERS_ALIASES#L115

anyhow good that you reviewed, thanks
ptal if you can

fossedihelm commented:
> > I don't have approval rights for this PR, so you need a root approver anyway :)
> 
> hmm what do i miss ? (just curious of course)
> 
> https://github.com/kubevirt/kubevirt/blob/5f866200e20553d7a37886ad9de99feea1fffece/OWNERS_ALIASES#L115

That is the list of the sig-compute approver:
```
  #
  # SIG Compute
  # Owns everything which is taking place on a node, for example
  # (but not limited to) groups, SELinux, node labels, …
  # And everything on the cluster level such as RBAC, controller, …
  #
  sig-compute-approvers:
    - jean-edouard
    - iholder101
    - fossedihelm
    ...
```

compute != root :)

fossedihelm commented:
> [APPROVALNOTIFIER] This PR is **NOT APPROVED**
> 
> This pull-request has been approved by: **Once this PR has been reviewed and has the lgtm label**, please assign [vasiliy-ul](https://github.com/vasiliy-ul) for approval. For more information see [the Code Review Process](https://github.com/kubevirt/community/blob/main/docs/add-merge-automation-to-your-repository.md?rgh-link-date=2025-05-19T07%3A50%3A03Z#example-workflow).
> 
> The full list of commands accepted by this bot can be found [here](https://prow.ci.kubevirt.io/command-help?repo=kubevirt%2Fkubevirt).
> 
> Needs approval from an approver in each of these files:
> * **[OWNERS](https://github.com/kubevirt/kubevirt/blob/main/OWNERS?rgh-link-date=2025-05-19T07%3A50%3A03Z)**
> 
> Approvers can indicate their approval by writing `/approve` in a comment Approvers can cancel approval by writing `/approve cancel` in a comment

https://github.com/kubevirt/kubevirt/pull/14623#issuecomment-2889982724

@oshoval Here you can see that this PR requires a root approver

fossedihelm commented:
/lgtm
From my POV. Thanks

enp0s3 commented:
/cc

enp0s3 reviewed (commented):
@oshoval Hi, in the `manager_test.go`, you will have to update the cleanup logic `defer os.Remove(socketPath)` with `os.RemoveAll(hookDir)` since now you have additional depth.

We will have to clean up this unit test suite, its awful that we are using real FS access and real gRPC

enp0s3 commented:
@oshoval Thank you for the effort! besides the previous comment LGTM

enp0s3 commented:
> > @oshoval Hi, in the `manager_test.go`, you will have to update the cleanup logic `defer os.Remove(socketPath)` with `os.RemoveAll(hookDir)` since now you have additional depth.
> > We will have to clean up this unit test suite, its awful that we are using real FS access and real gRPC
> 
> we have already
> 
> ```
> 		AfterEach(func() {
> 			os.RemoveAll(socketDir)
> 		})
> ```
> 
> which deletes the temp dir that is created on the BeforeEach, it will take care about the additional depth, each test deletes it's sockets already, same as was done before

@oshoval Wow, its even worse :) thanks for the heads up! Not clear then why defer close is used inside the test.

/approve
